main/src/optparse.hxx

```diff
@@ -0,0 +1,367 @@
+///    // positional arguments:
+///    myOpts.fArgs = opts.GetArgs();
+///
+///    return myOpts;
+/// }
+/// ~~~
+///
+/// ## Additional Notes
+/// If all the short flags you pass (those starting with a single `-`) are 1 character long, the parser will accept
+/// grouped flags like "-abc" as equivalent to "-a -b -c". The last flag in the group may also accept an argument, in
+/// which case "-abc foo" will count as "-a -b -c foo" where "foo" is the argument to "-c".
+///
+/// The string "--" is treated as the positional argument separator: all strings after it will be treated as positional
+/// arguments even if they start with "-".
+///
+/// \author Giacomo Parolini <giacomo.parolini@cern.ch>
+/// \date 2025-10-09
+
+#ifndef ROOT_OptParse
+#define ROOT_OptParse
+
+#include <algorithm>
+#include <charconv>
+#include <cstring>
+#include <iostream>
+#include <limits>
+#include <optional>
+#include <sstream>
+#include <stdexcept>
+#include <string>
+#include <string_view>
+#include <vector>
+
+namespace ROOT {
+
+class RCmdLineOpts {
+public:
+   enum class EFlagType {
+      kBoolean,
+      kWithArg
+   };
+
+   struct RFlag {
+      std::string fName;
+      std::string fValue;
+   };
+
+private:
+   std::vector<RFlag> fFlags;
+   std::vector<std::string> fArgs;
+   // If true, many short flags may be grouped: "-abc" == "-a -b -c".
+   // This is automatically true if all short flags given are 1 character long, otherwise it's false.
+   bool fAllowFlagGrouping = true;
+
+   struct RExpectedFlag {
+      EFlagType fFlagType = EFlagType::kBoolean;
+      std::string fName;
+      std::string fHelp;
+      // If >= 0, this flag is an alias of the RExpectedFlag at index fAlias.
+      int fAlias = -1;
+      bool fShort = false;
+
+      std::string AsStr() const { return std::string(fShort ? "-" : "--") + fName; }
+   };
+   std::vector<RExpectedFlag> fExpectedFlags;
+   std::vector<std::string> fErrors;
+
+   const RExpectedFlag *GetExpectedFlag(std::string_view name) const
+   {
+      for (const auto &flag : fExpectedFlags) {
+         if (flag.fName == name)
+            return &flag;
+      }
+      return nullptr;
+   }
+
+public:
+   const std::vector<std::string> &GetErrors() const { return fErrors; }
+   const std::vector<std::string> &GetArgs() const { return fArgs; }
+   const std::vector<RFlag> &GetFlags() const { return fFlags; }
+
+   /// Conveniency method to print any errors to `stream`.
+   /// \return true if any error was printed
+   bool ReportErrors(std::ostream &stream = std::cerr) const
+   {
+      for (const auto &err : fErrors)
+         stream << err << "\n";
+      return !fErrors.empty();
+   }
+
+   void AddFlag(std::initializer_list<std::string_view> aliases, EFlagType type = EFlagType::kBoolean,
+                std::string_view help = "")
+   {
+      int aliasIdx = -1;
+      for (auto f : aliases) {
```

Comment:
Perhaps: check that the alias is neither `-` nor contains a space (and possibly other funny stuff, newline, NULL, etc). Maybe enough to say in the documentation what is allowed.